### PR TITLE
Refs [TASK][28541]:Get data from API

### PR DESCRIPTION
**Pull request by dinhthuyha**
## Related Tickets
- [#28541: Get data from API](https://edu-redmine.sun-asterisk.vn/issues/28541)

## WHAT
- Get data from API
## HOW
-Get data from API
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [x] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [x] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseDetailGenre.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.sunasterisk.movieintroduction.data.model
+
+class ResponseDetailGenre(
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseDetailGenre.kt]:
```suggestion
data class DetailGenreResponse(
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseDetailGenre.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    var poster_path: String,
+    var id: String,
+    var title: String,
+    var genre_ids: List<Int>,
+    var vote_average: Int,
+    var release_date: String
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseDetailGenre.kt]: Đặt tên thuộc tính theo camelCase

File: app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseGenre.kt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+package com.sunasterisk.movieintroduction.data.model
+
+class ResponseGenre( var id: Int, var name: String) {
+}
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseGenre.kt]: format code, ưu tiên val hơn var 
Những thuộc tính nào có nhu cầu thay đổi về sau thì mới để var
nếu mặc định khởi tạo thông qua constructor thì thôi 
```suggestion
class GenreResponse(val id: Int, val name: String) {
}
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+class MovieRepository private constructor(private var remoteGenre: MovieDataSource.Remote ): MovieDataSource.Remote{
+    override fun getAllGenres(): List<ResponseGenre> {
+        return remoteGenre.getAllGenres()
+    }
+    companion object{
+        private var INSTANCE: MovieRepository? = null
+        fun getInstance(remoteDataSource: MovieDataSource.Remote) =
+            INSTANCE ?: synchronized(this) {
+                INSTANCE ?: MovieRepository(remoteDataSource)
+            }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt]:
```suggestion
            }.also { INSTANCE = it }
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.sunasterisk.movieintroduction.data.source
+
+import com.sunasterisk.movieintroduction.data.model.ResponseGenre
+import com.sunasterisk.movieintroduction.data.source.remote.HomeData
+
+class MovieRepository private constructor(private var remoteGenre: MovieDataSource.Remote ): MovieDataSource.Remote{
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt]: Chú ý class header 
https://kotlinlang.org/docs/reference/coding-conventions.html#class-header-formatting
```suggestion
class MovieRepository private constructor(
    private val remoteGenre: MovieDataSource.Remote 
): MovieDataSource.Remote{
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    override fun getAllGenres(): List<ResponseGenre> {
+        return remoteGenre.getAllGenres()
+    }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt]:
```suggestion
    override fun getAllGenres(): List<ResponseGenre> =
        remoteGenre.getAllGenres()
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeData.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import com.sunasterisk.movieintroduction.data.source.MovieDataSource
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.BufferedReader
+import java.io.IOException
+import java.io.InputStreamReader
+import java.net.MalformedURLException
+import java.net.URL
+
+object HomeData {
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeData.kt]: thêm object vào để làm gi nhỉ?

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeData.kt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    class homeTask() : AsyncTask<String?, Void?, List<ResponseGenre>>(), MovieDataSource.Remote {
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeData.kt]: Tên class đặt theo mẫu HomeTask

Comment (huytq-2422) [reply app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeData.kt]: sai mô hình rồi, sao AsyncTask lại kế thừa Remote

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/MovieRemoteDataSource.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import com.sunasterisk.movieintroduction.data.model.ResponseGenre
+import com.sunasterisk.movieintroduction.data.source.MovieDataSource
+
+class MovieRemoteDataSource() : MovieDataSource.Remote {
+    var key: String = BuildConfig.API_KEY
+    override fun getAllGenres(): List<ResponseGenre> {
+        return HomeData.homeTask()
+            .execute("https://api.themoviedb.org/3/genre/movie/list?api_key=45519d62d8e3f306ba1903546f926b0c&language=en-US")
+            .get()
+    }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/MovieRemoteDataSource.kt]: lộ key, sai mô hình

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -1,3 +1,5 @@
 <resources>
     <string name="app_name">Movie Introduction</string>
+    <!-- TODO: Remove or change this placeholder text -->
+    <string name="hello_blank_fragment">Hello blank fragment</string>
```
Comment (huytq-2422) [inline app/src/main/res/values/strings.xml]: remove

File: gradle.properties [hunk 11]
```diff
@@ -19,3 +19,4 @@ android.useAndroidX=true
 android.enableJetifier=true
 # Kotlin code style for this project: "official" or "obsolete":
 kotlin.code.style=official
+API_KEY=45519d62d8e3f306ba1903546f926b0c
```
Comment (huytq-2422) [inline gradle.properties]: lộ key

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            }
+
+        } catch (e: MalformedURLException) {
+            e.printStackTrace()
+        } catch (e: IOException) {
+            e.printStackTrace()
+        } catch (e: JSONException) {
+            e.printStackTrace()
+        }
+        Log.d("AA", "doInBackground: ${list.size}")
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt]: xóa log

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+                list.add(ResponseGenre(id, name))
+            }
+
+        } catch (e: MalformedURLException) {
+            e.printStackTrace()
+        } catch (e: IOException) {
+            e.printStackTrace()
+        } catch (e: JSONException) {
+            e.printStackTrace()
+        }
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt]: Xử lý exception

Comment (dinhthuyha) [reply app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt]: nên đóng kết nối hay như nào ạ

Comment (aug16vcc) [reply app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt]:
```suggestion
        } catch (e: Exception) {
            when (e) { /* type check */ }
        } finally {
            connection.close()
        }
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        Log.d("AA", "doInBackground: ${list.size}")
+        return list
+    }
+
+
+
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt]: Bỏ dòng trống

File: app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseGenre.kt [hunk 15]
```diff
@@ -0,0 +1,4 @@
+package com.sunasterisk.movieintroduction.data.model
+
+data class ResponseGenre( val id: Int, val name: String) {
+}
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/model/ResponseGenre.kt:4]:
```suggestion
data class ResponseGenre(val id: Int, val name: String)
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+): MovieDataSource.Remote{
+    override fun getAllGenres(): List<ResponseGenre> =
+        remoteGenre.getAllGenres()
+
+    companion object{
+        private var INSTANCE: MovieRepository? = null
+        fun getInstance(remoteDataSource: MovieDataSource.Remote) =
+            INSTANCE ?: synchronized(this) {
+                INSTANCE ?: MovieRepository(remoteDataSource)
+            }.also { INSTANCE=it }
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/MovieRepository.kt:17]:
```suggestion
            }.also { INSTANCE = it }
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+import com.sunasterisk.movieintroduction.data.source.MovieDataSource
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.BufferedReader
+import java.io.IOException
+import java.io.InputStreamReader
+import java.net.MalformedURLException
+import java.net.URL
+
+class HomeTask() : AsyncTask<String?, Void?, List<ResponseGenre>>() {
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt:17]:
```suggestion
class HomeTask() : AsyncTask<String?, Unit, List<ResponseGenre>>() {
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    var list = mutableListOf<ResponseGenre>()
+    var genre_ids = mutableListOf<Int>()
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt:19]:
```suggestion
    private val list = mutableListOf<ResponseGenre>()
    private val genreIds = mutableListOf<Int>()
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    override fun doInBackground(vararg p0: String?): List<ResponseGenre> {
+        var content = StringBuffer()
+        try {
+            var url = URL(p0[0])
+            var inputStreamReader = InputStreamReader(url.openConnection().getInputStream())
+            val bufferedReader = BufferedReader(inputStreamReader)
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt]:
```suggestion
            val url = URL(p0[0])
            val inputStreamReader = InputStreamReader(url.openConnection().getInputStream())
            val bufferedReader = BufferedReader(inputStreamReader)
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+            var line: String? = ""
+            while (bufferedReader.readLine().also { line = it } != null) {
+                content.append(line)
+            }
+            var jsonObject = JSONObject(content.toString())
+            var genres = jsonObject.getJSONArray("genres")
+            for (i in 0 until genres.length()) {
+                var ob = genres.getJSONObject(i)
+                var id = ob.getInt("id")
+                var name = ob.getString("name")
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt:37]:
```suggestion
                val ob = genres.getJSONObject(i)
                val id = ob.getInt("id")
                val name = ob.getString("name")
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+        try {
+            var url = URL(p0[0])
+            var inputStreamReader = InputStreamReader(url.openConnection().getInputStream())
+            val bufferedReader = BufferedReader(inputStreamReader)
+            var line: String? = ""
+            while (bufferedReader.readLine().also { line = it } != null) {
+                content.append(line)
+            }
+            var jsonObject = JSONObject(content.toString())
+            var genres = jsonObject.getJSONArray("genres")
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/HomeTask.kt:33]:
```suggestion
            val jsonObject = JSONObject(content.toString())
            val genres = jsonObject.getJSONArray("genres")
```

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/MovieRemoteDataSource.kt [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+import com.sunasterisk.movieintroduction.BuildConfig
+import com.sunasterisk.movieintroduction.data.model.ResponseGenre
+import com.sunasterisk.movieintroduction.data.source.MovieDataSource
+
+class MovieRemoteDataSource() : MovieDataSource.Remote {
+    var key: String = BuildConfig.API_KEY
+    override fun getAllGenres(): List<ResponseGenre> {
+        return HomeTask()
+            .execute("https://api.themoviedb.org/3/genre/movie/list?api_key=$key&language=en-US").get()
+
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/MovieRemoteDataSource.kt:12]: Dòng trống

File: app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/MovieRemoteDataSource.kt [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+import com.sunasterisk.movieintroduction.BuildConfig
+import com.sunasterisk.movieintroduction.data.model.ResponseGenre
+import com.sunasterisk.movieintroduction.data.source.MovieDataSource
+
+class MovieRemoteDataSource() : MovieDataSource.Remote {
+    var key: String = BuildConfig.API_KEY
+    override fun getAllGenres(): List<ResponseGenre> {
+        return HomeTask()
+            .execute("https://api.themoviedb.org/3/genre/movie/list?api_key=$key&language=en-US").get()
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/data/source/remote/MovieRemoteDataSource.kt:11]: `hard coding` chuyển sang 1 `const val`

File: app/src/main/java/com/sunasterisk/movieintroduction/ui/main/MainActivity.kt [hunk 24]
```diff
@@ -1,12 +1,15 @@
...
 import android.os.Bundle
+import com.sunasterisk.movieintroduction.R
+import com.sunasterisk.movieintroduction.data.source.remote.HomeData
 
 class MainActivity : AppCompatActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/movieintroduction/ui/main/MainActivity.kt:12]: Blank line

File: app/src/main/res/values/strings.xml [hunk 25]
```diff
@@ -1,3 +1,4 @@
 <resources>
     <string name="app_name">Movie Introduction</string>
+    <!-- TODO: Remove or change this placeholder text -->
```
Comment (aug16vcc) [inline app/src/main/res/values/strings.xml:3]: `remove`

**Review by aug16vcc — CHANGES_REQUESTED** [review]
Còn sử dụng java code style nhiều. Cải thiện nhé

**Pull request closed without merging**